Plug & Heroku Deployment Post

notactuallypagemcconnell opened the pull request:
This is a WIP, definitely want to review and walk through the steps again to ensure it all works copy-paste etc. Just wanted to get discussion started!

notactuallypagemcconnell commented on _posts/2019-02-04-putting-a-plug-app-on-heroku.md:
```diff
@@ -0,0 +1,98 @@
+---
+
+In our [previous_post](TODO_LINK) we built a very simple Portfolio app with `Plut.Router`.
+In this (brief) post, we will show how to get things up and running on [Heroku](http://heroku.com) for a vanilla `Plug` app.
```
Is using `vanilla` to describe the app here difficult/weird to translate? Could pick something better. Upon further thought I imagine it probably sounds silly in other languages or might confuse translators who haven't heard the term used that way before.

notactuallypagemcconnell commented on _posts/2019-02-04-putting-a-plug-app-on-heroku.md:
```diff
@@ -0,0 +1,98 @@
+
+In our [previous_post](TODO_LINK) we built a very simple Portfolio app with `Plut.Router`.
+In this (brief) post, we will show how to get things up and running on [Heroku](http://heroku.com) for a vanilla `Plug` app.
+It really is quite easy, but today when I set out to do it, the resources I found either focused on Phoenix or had a few holes, so I figured I would write up a simple one here.
```
I'm meh on this line, should I axe it?

notactuallypagemcconnell commented on _posts/2019-02-04-putting-a-plug-app-on-heroku.md:
````diff
@@ -0,0 +1,98 @@
+Now, we want to set a few configs for heroku:
+
+```
+$ heroku config:set MIX_ENV=prod
````
redundant - need to remove the mix env thing here its done above.

doomspork commented on _posts/2019-02-04-putting-a-plug-app-on-heroku.md:
```diff
@@ -0,0 +1,98 @@
+
+Thats it!
+
+Now, the reason Heroku is so easy is because they have a _lot_ of built in tooling to make deploying your app easy.
```
The joys of using Heroku is that they handle a lot of the work for you with awesome builtin tooling.
One of those features is the `buildpacks` which include instructions for building, in our case compiling, our application code.
For Elixir we'll need to add a specialized `buildpack` that knows how to fetch Elixir dependencies and build that code.

...

doomspork commented on _posts/2019-02-04-putting-a-plug-app-on-heroku.md:
```diff
@@ -0,0 +1,98 @@
+
+Now, the reason Heroku is so easy is because they have a _lot_ of built in tooling to make deploying your app easy.
+In the case of Elixir, we need a buildpack that will handle our compilation and to run the app and get the proper dependencies on our server.
+To get any further, you will need a Heroku account and to have their toolbelt installed.
```
I think we should cover setting up your Heroku account earlier and almost as a footnote, it's not the focus of this post _really_.

doomspork commented on _posts/2019-02-04-putting-a-plug-app-on-heroku.md:
````diff
@@ -0,0 +1,98 @@
+heroku buildpacks:set https://github.com/HashNuke/heroku-buildpack-elixir
+```
+
+Now, we need to set our env to prod on heroku:
````
Now we need to setup our environment variables.
For our application we only need to worry about one: `MIX_ENV`.

doomspork commented on _posts/2019-02-04-putting-a-plug-app-on-heroku.md:
````diff
@@ -0,0 +1,98 @@
+heroku config:set MIX_ENV=prod
+```
+
+We also need to give our buildpack some config so it knows what our Elixir/Erlang environments are like.
````
Let's move this up just after we add the buildpack.

doomspork commented on _posts/2019-02-04-putting-a-plug-app-on-heroku.md:
````diff
@@ -0,0 +1,98 @@
+Now that we've set this, let's change our application initialization to use the configured port from the env:
+
+```
+# lib/router_example/application.ex
````
Let's not start this tread, mention the file in the previous sentence if you need to call attention to the name 👍

doomspork commented on _posts/2019-02-04-putting-a-plug-app-on-heroku.md:
````diff
@@ -0,0 +1,98 @@
+```
+# lib/router_example/application.ex
+  
+  # ....
````
This is not necessary.  The file is short enough to be included in its entirety.

doomspork commented on _posts/2019-02-04-putting-a-plug-app-on-heroku.md:
```diff
@@ -0,0 +1,98 @@
+    |> String.to_integer()
+  end
+
+  # ....
```
Same here. We're removing `end` and that's about it.

doomspork commented on _posts/2019-02-04-putting-a-plug-app-on-heroku.md:
````diff
@@ -0,0 +1,98 @@
+$ heroku config:set PORT=4001
+```
+
+Now that we've set this, let's change our application initialization to use the configured port from the env:
````
This note can go away because we're not configuring a port. Instead we should mention that we need to update our `lib/router_example/application.ex` file to fetch the port from the ENV.

doomspork commented on _posts/2019-02-04-putting-a-plug-app-on-heroku.md:
````diff
@@ -0,0 +1,98 @@
+  # ....
+```
+
+With this, we add a function to get our port.
````
I think there's a few extra commas here but I think we can reward this given some previous comments:

With the addition of `application_port/0` we're now setup to check the environment variables for `PORT` provided by Heroku and when not present use our default.

doomspork commented on _posts/2019-02-04-putting-a-plug-app-on-heroku.md:
```diff
@@ -0,0 +1,98 @@
+With this, we add a function to get our port.
+We either get it from the environment, or default to port 4001.
+
+Now we can get it up in the real world:
```
Maybe we can rephrase this?  "get it up in the real world" doesn't make much sense in English and will be difficult for those who don't speak it.

doomspork commented on _posts/2019-02-04-putting-a-plug-app-on-heroku.md:
````diff
@@ -0,0 +1,98 @@
+$ git push heroku master && heroku open
+```
+
+And after that push succeeds, we can see our portfolio site live and working in the real world.
````
Once the push, and build, have completed we should see our portfolio site live on Heroku!

doomspork commented on _posts/2019-02-04-putting-a-plug-app-on-heroku.md:
```diff
@@ -0,0 +1,98 @@
+
+And after that push succeeds, we can see our portfolio site live and working in the real world.
+
+We don't go over the granularity of test/dev/prod configuration here.
```
This entire paragraph can go, it doesn't really fit with the rest of the lesson.

doomspork commented on _posts/2019-02-04-putting-a-plug-app-on-heroku.md:
```diff
@@ -0,0 +1,98 @@
+categories: general
+date: 2019-02-03
+layout: post
+title: Getting a Plug App Working With Heroku
```
Deploying out `Plug.Router` application to Heroku

doomspork commented on _posts/2019-02-04-putting-a-plug-app-on-heroku.md:
```diff
@@ -0,0 +1,98 @@
+---
+author: Bobby Grayson
+categories: general
+date: 2019-02-03
```
We'll want to change this date to whenever we release the post.

notactuallypagemcconnell commented:
@doomspork I believe I handled all feedback. I removed the entirety of the port config in `application.ex` because it was unneeded to just get it working. Just pushed and going to re-read in a bit once I've not just been staring so I have a fresh look.

doomspork commented:
Sounds good @notactuallypagemcconnell, I'll re-read it this weekend. I talked with @SophieDeBenedetto about how to stagger these blogs now that we have three on deck. I'm going to merge and announce #1710 today. That gives us the weekend for others to review this post, we'll target a Monday post of this. Followed by #1713 later in the week.

notactuallypagemcconnell commented:
Sounds good @doomspork!

It looks like some outdated comments are still showing -- I've updated the date and the paragraphs you suggested I have added too but they keep showing. Odd. excited to hear more feedback from others, I'm impressed with how quick/easy this is.

doomspork commented:
I'm not sure what's up with GH lately. Having to go back and manually resolve comments is a PITA.
